subsys/net/lib/coap/coap_client.c

```diff
@@ -0,0 +1,716 @@
+{
+	if (client->coap_request->cb) {
+		client->coap_request->cb(error_code, 0, NULL, 0, true,
+					 client->coap_request->user_data);
+	}
+}
+
+static int send_ack(struct coap_client *client, const struct coap_packet *req,
+		    uint8_t response_code)
+{
+	int ret;
+
+	ret = coap_ack_init(&client->request, req, client->send_buf, MAX_COAP_MSG_LEN,
+			    response_code);
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize CoAP ACK-message");
+		return ret;
+	}
+
+	ret = send_request(client->fd, client->request.data, client->request.offset, 0,
+			   &client->address, client->socklen);
+	if (ret < 0) {
+		LOG_ERR("Error sending a CoAP ACK-message");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int send_reset(struct coap_client *client, const struct coap_packet *req,
+		      uint8_t response_code)
+{
+	int ret;
+	uint16_t id;
+	uint8_t token[COAP_TOKEN_MAX_LEN];
+	uint8_t tkl;
+
+	id = coap_header_get_id(req);
+	tkl = response_code ? coap_header_get_token(req, token) : 0;
+	ret = coap_packet_init(&client->request, client->send_buf, MAX_COAP_MSG_LEN, COAP_VERSION,
+			       COAP_TYPE_RESET, tkl, token, response_code, id);
+
+	if (ret < 0) {
+		LOG_ERR("Error creating CoAP reset message");
+		return ret;
+	}
+
+	ret = send_request(client->fd, client->request.data, client->request.offset, 0,
+			   &client->address, client->socklen);
+	if (ret < 0) {
+		LOG_ERR("Error sending CoAP reset message");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int handle_response(struct coap_client *client, const struct coap_packet *response)
+{
+	int ret = 0;
+	int response_type;
+	int block_option;
+	int block_num;
+	bool blockwise_transfer = false;
+	bool last_block = false;
+
+	/* Handle different types, ACK might be separate or piggybacked
+	 * CON and NCON contains a separate response, CON needs an empty response
+	 * CON request results as ACK and possibly separate CON or NCON response
+	 * NCON request results only as a separate CON or NCON message as there is no ACK
+	 * With RESET, just drop gloves and call the callback.
+	 */
+	response_type = coap_header_get_type(response);
+
+	/* Reset and Ack need to match the message ID with request */
+	if ((response_type == COAP_TYPE_ACK || response_type == COAP_TYPE_RESET) &&
+	     coap_header_get_id(response) != client->pending.id)  {
+		LOG_ERR("Unexpected ACK or Reset");
+		return -EFAULT;
+	} else if (response_type == COAP_TYPE_RESET) {
+		coap_pending_clear(&client->pending);
+	}
+
+	/* CON, NON_CON and piggybacked ACK need to match the token with original request */
+	uint16_t payload_len;
+	uint8_t response_code = coap_header_get_code(response);
+	const uint8_t *payload = coap_packet_get_payload(response, &payload_len);
+
+	/* Separate response */
+	if (payload_len == 0 && response_type == COAP_TYPE_ACK &&
+	    response_code == COAP_CODE_EMPTY) {
+		/* Clear the pending, poll uses now the separate timeout for the response. */
+		coap_pending_clear(&client->pending);
+		return 1;
```

Comment:
It would be useful for the user of this library to get a callback when an ACK is received in response to a CON request that carries no additional payload (such as a PUT or POST).  Otherwise, the user does not know if the server accepted the request.

Reply:
But it might be confusing for the user if the server responds with ACK, but doesn't send response ever if the separate response takes too long.

Reply:
OK, sounds good.  It is important for the user to get a callback if the separate response never arrives, and retries are exhausted.